cmd/store.go

```diff
@@ -23,52 +24,46 @@ func (s *store) Get(IP string) (string, error) {
 	if role, ok := s.rolesByIP[IP]; ok {
 		return role, nil
 	}
+	log.Warnf("Using fallback role for IP %s", IP)
 	if s.defaultRole != "" {
-		log.Warnf("Using fallback role for IP %s", IP)
 		return s.defaultRole, nil
 	}
 	return "", fmt.Errorf("Unable to find role for IP %s", IP)
 }
 
 // OnAdd is called when a pod is added.
 func (s *store) OnAdd(obj interface{}) {
-	if pod, ok := obj.(*api.Pod); ok {
-		if pod.Status.PodIP != "" {
-			if role, ok := pod.Annotations[s.iamRoleKey]; ok {
-				s.mutex.Lock()
-				s.rolesByIP[pod.Status.PodIP] = role
-				s.mutex.Unlock()
-			}
+	pod := obj.(*api.Pod)
+	if pod.Status.PodIP != "" {
+		if role, ok := pod.Annotations[s.iamRoleKey]; ok {
+			s.mutex.Lock()
+			s.rolesByIP[pod.Status.PodIP] = role
+			s.mutex.Unlock()
 		}
 	}
 }
 
 // OnUpdate is called when a pod is modified.
 func (s *store) OnUpdate(oldObj, newObj interface{}) {
-	oldPod, okOld := oldObj.(*api.Pod)
-	newPod, okNew := newObj.(*api.Pod)
+	oldPod := oldObj.(*api.Pod)
+	newPod := newObj.(*api.Pod)
 
-	// Validate that the objects are good
-	if okOld && okNew {
-		if oldPod.Status.PodIP != newPod.Status.PodIP {
-			s.OnDelete(oldPod)
-			s.OnAdd(newPod)
-		}
-	} else if okNew {
-		s.OnAdd(newPod)
-	} else if okOld {
+	if oldPod.Status.PodIP != newPod.Status.PodIP {
 		s.OnDelete(oldPod)
+		s.OnAdd(newPod)
 	}
 }
 
 // OnDelete is called when a pod is deleted.
 func (s *store) OnDelete(obj interface{}) {
-	if pod, ok := obj.(*api.Pod); ok {
-		if pod.Status.PodIP != "" {
-			s.mutex.Lock()
-			delete(s.rolesByIP, pod.Status.PodIP)
-			s.mutex.Unlock()
-		}
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		pod = obj.(kcache.DeletedFinalStateUnknown).Obj.(*api.Pod)
```

Comment:
Thanks, seems to make sense given the comment on https://github.com/kubernetes/kubernetes/blob/8fd414537b5143ab039cb910590237cabf4af783/pkg/client/cache/controller.go#L146. Good find 👍